File: ports/ruby/test/stream_safe_violation_test.rb
# frozen_string_literal: true

require_relative "test_helper"

class StreamSafeViolationTest < Minitest::Test
  include RubyPortTestHelpers

  StreamSafe = UnicodeRuby::Security::Form::StreamSafeViolation
  Policy = UnicodeRuby::Security::Policy
  Family = UnicodeRuby::Security::Calculus::Family

  ACUTE = 0x0301 # COMBINING ACUTE ACCENT, CCC = 230 (a non-starter).

  # Reason code the detector composes for its single sub-threat.
  def reason_code_for(verdict)
    tag = verdict.classify.tag
    return nil if tag.nil?

    Policy.reason_code(Family::STREAM_SAFE_VIOLATION, tag)
  end

  # Every case in the shared detector fixture must round-trip through detect:
  # its composed reason code matches the fixture's required findings, and a
  # clear case emits no stream-safe reason.
  def test_shared_fixture_cases
    fixture = fixture_json("detectors", "stream_safe_violation.json")
    assert_equal 1, fixture.fetch("schema")
    assert_equal "stream-safe-violation", fixture.fetch("family")

    fixture.fetch("cases").each do |case_data|
      name = case_data.fetch("name")
      verdict = StreamSafe.detect(case_data.fetch("input"))
      code = reason_code_for(verdict)
      required = case_data.fetch("required_findings")

      if required.empty?
        assert verdict.classify.clear?, "#{name}: expected clear"
        assert_nil code, "#{name}: unexpected finding #{code.inspect}"
      else
        refute verdict.classify.clear?, "#{name}: expected hazard"
        assert_equal required, [code], name
      end
    end
  end

  # Build "a" followed by n combining acute accents.
  def a_plus_marks(count)
    [0x61] + ([ACUTE] * count)
  end

  # Exactly 30 marks after a starter is the boundary case — stays clear under
  # strict `>`.
  def test_thirty_marks_boundary_clear
    verdict = StreamSafe.detect(a_plus_marks(30))
    assert verdict.classify.clear?
    assert_nil verdict.classify.tag
    assert_equal 30, verdict.max_run_len
    assert_equal 0, verdict.overrun_count
    assert_equal 30, verdict.total_non_starters
  end

  # 31 marks after a starter fires StreamSafeOverrun with base_pos 1, run_len
  # 31, positions [1].
  def test_thirtyone_marks_overrun
    verdict = StreamSafe.detect(a_plus_marks(31))
    refute verdict.classify.clear?
    assert_equal "StreamSafeOverrun", verdict.classify.tag
    assert_equal [1], verdict.classify.implicated_positions
    assert_equal :stream_safe_overrun, verdict.classify.sub.kind
    assert_equal 1, verdict.classify.sub.base_pos
    assert_equal 31, verdict.classify.sub.run_len
    assert_equal 31, verdict.max_run_len
    assert_equal 1, verdict.overrun_count
    assert_equal 31, verdict.total_non_starters
    assert_equal "unicode.security.F.stream-safe-violation.StreamSafeOverrun",
                 reason_code_for(verdict)
  end

  # Empty and pure-ASCII inputs are clear with zeroed summaries.
  def test_empty_and_ascii_clear
    empty = StreamSafe.detect([])
    assert empty.classify.clear?
    assert_equal 0, empty.max_run_len
    assert_equal 0, empty.total_non_starters

    ascii = StreamSafe.detect([0x48, 0x65, 0x6C, 0x6C, 0x6F])
    assert ascii.classify.clear?
    assert_equal 0, ascii.max_run_len
  end

  # A bare non-starter run that opens at index 0 records its start as 0.
  def test_bare_mark_run_starts_at_zero
    verdict = StreamSafe.detect([ACUTE] * 31)
    assert_equal "StreamSafeOverrun", verdict.classify.tag
    assert_equal [0], verdict.classify.implicated_positions
    assert_equal 31, verdict.max_run_len
    assert_equal 31, verdict.total_non_starters
  end

  # The first overrun wins: a short run before a long run does not shadow it,
  # and the reported base_pos is the long run's start.
  def test_first_overrun_reports_long_run_start
    input = a_plus_marks(5) + [0x62] + ([ACUTE] * 31)
    verdict = StreamSafe.detect(input)
    assert_equal "StreamSafeOverrun", verdict.classify.tag
    assert_equal [7], verdict.classify.implicated_positions
    assert_equal 31, verdict.max_run_len
    assert_equal 1, verdict.overrun_count
    assert_equal 36, verdict.total_non_starters
  end

  # Two separate runs, each under the limit, stay clear but are both summed.
  def test_two_short_runs_clear_totals_summed
    input = a_plus_marks(30) + [0x62] + ([ACUTE] * 30)
    verdict = StreamSafe.detect(input)
    assert verdict.classify.clear?
    assert_equal 30, verdict.max_run_len
    assert_equal 0, verdict.overrun_count
    assert_equal 60, verdict.total_non_starters
  end
end

File: ports/elixir/test/stream_safe_violation_test.exs
defmodule UnicodeSecurity.Form.StreamSafeViolationTest do
  use ExUnit.Case, async: false
  alias UnicodeSecurity.Form.StreamSafeViolation
  alias UnicodeSecurity.Policy
  import UnicodeSecurity.TestHelpers

  @acute 0x0301

  defp a_plus_marks(n), do: [0x61 | List.duplicate(@acute, n)]

  test "shared fixture through the form-scan policy path" do
    fixture = fixture_json(Path.join("detectors", "stream_safe_violation.json"))

    Enum.each(fixture["cases"], fn case_data ->
      verdict = Policy.scan_forms("gateway-header", "observe", case_data["input"])
      codes = codes(verdict)

      Enum.each(case_data["required_findings"], fn required ->
        assert required in codes, "#{fixture["family"]}/#{case_data["name"]}"
      end)

      if case_data["required_findings"] == [] do
        needle = "." <> fixture["family"] <> "."

        Enum.each(codes, fn code ->
          refute String.contains?(code, needle), "#{fixture["family"]}/#{case_data["name"]}"
        end)
      end
    end)
  end

  test "reason code matches the fixture required finding" do
    assert Policy.reason_code(:stream_safe_violation, "StreamSafeOverrun") ==
             "unicode.security.F.stream-safe-violation.StreamSafeOverrun"
  end

  test "limit is 30" do
    assert StreamSafeViolation.stream_safe_limit() == 30
  end

  test "empty input is clear" do
    v = StreamSafeViolation.detect([])
    assert v.classify.kind == :clear
    assert v.sub == nil
    assert v.positions == []
    assert v.max_run_len == 0
    assert v.overrun_count == 0
    assert v.total_non_starters == 0
  end

  test "pure ASCII is clear" do
    v = StreamSafeViolation.detect([0x48, 0x65, 0x6C, 0x6C, 0x6F])
    assert v.classify.kind == :clear
    assert v.max_run_len == 0
    assert v.total_non_starters == 0
  end

  test "one combining mark is clear" do
    v = StreamSafeViolation.detect([0x61, @acute])
    assert v.classify.kind == :clear
    assert v.max_run_len == 1
    assert v.overrun_count == 0
    assert v.total_non_starters == 1
  end

  test "exactly 30 marks is the boundary — stays clear under strict >" do
    v = StreamSafeViolation.detect(a_plus_marks(30))
    assert v.classify.kind == :clear
    assert v.sub == nil
    assert v.max_run_len == 30
    assert v.overrun_count == 0
    assert v.total_non_starters == 30
  end

  test "31 marks fires StreamSafeOverrun at base_pos 1" do
    v = StreamSafeViolation.detect(a_plus_marks(31))
    assert v.classify.kind == :hazard
    assert v.sub == "StreamSafeOverrun"
    assert v.positions == [1]
    assert v.classify.sub == %{tag: "StreamSafeOverrun", base_pos: 1, run_len: 31}
    assert v.max_run_len == 31
    assert v.overrun_count == 1
    assert v.total_non_starters == 31
  end

  test "a bare non-starter run records its start at index 0" do
    v = StreamSafeViolation.detect(List.duplicate(@acute, 31))
    assert v.sub == "StreamSafeOverrun"
    assert v.positions == [0]
    assert v.max_run_len == 31
    assert v.total_non_starters == 31
  end

  test "two short runs stay clear but totals sum" do
    input = a_plus_marks(30) ++ [0x62] ++ List.duplicate(@acute, 30)
    v = StreamSafeViolation.detect(input)
    assert v.classify.kind == :clear
    assert v.max_run_len == 30
    assert v.overrun_count == 0
    assert v.total_non_starters == 60
  end

  test "the first overrun wins and reports the long run's start" do
    input = a_plus_marks(5) ++ [0x62] ++ List.duplicate(@acute, 31)
    v = StreamSafeViolation.detect(input)
    assert v.sub == "StreamSafeOverrun"
    assert v.positions == [7]
    assert v.max_run_len == 31
    assert v.overrun_count == 1
    assert v.total_non_starters == 36
  end
end
